File: tools/market_map/validation_engine.py
#!/usr/bin/env python3
"""
validation_engine.py — research-overfit promotion gate (the "Omitted Strategies" #2): make overfit control
a MACHINE-READABLE gate, not a research note. Provides:
  purged_kfold(n, k, embargo)  purged + embargoed time-series CV splits (no serial-correlation leakage).
  pbo_cscv(M, S)               Probability of Backtest Overfitting via Combinatorially-Symmetric CV
                               (Bailey, Borwein, Lopez de Prado, Zhu 2017). Pure noise -> ~0.5; real edge -> low.
  promotion_gate(dsr, pbo)     deployable iff DSR >= min_dsr AND PBO <= max_pbo (Deflated Sharpe lives in
                               rank_engine.deflated_sharpe). Returns {deployable, reasons}.

Verified-engine pattern: authoritative; validation_engine.js is locked to tools/validation_golden.json.
Pure-stdlib. Research only.
"""
import json, math, os, itertools

__all__ = ["purged_kfold", "pbo_cscv", "promotion_gate"]


def purged_kfold(n, k, embargo=0):
    """Purged + embargoed k-fold splits for time series: each test fold is a CONTIGUOUS block; the train
    set excludes the test block AND `embargo` observations on each side. Returns [(train_idx, test_idx)]."""
    if k < 2 or n < k:
        return []
    fold = n // k; out = []
    for i in range(k):
        lo = i * fold; hi = n if i == k - 1 else (i + 1) * fold
        test = list(range(lo, hi))
        elo = max(0, lo - embargo); ehi = min(n, hi + embargo)
        train = [j for j in range(n) if j < elo or j >= ehi]
        out.append((train, test))
    return out


def _mean(xs):
    return sum(xs) / len(xs) if xs else 0.0


def pbo_cscv(M, S=8):
    """Probability of Backtest Overfitting (CSCV). M: T x N performance matrix (rows=periods,
    cols=strategy candidates). Partition rows into S even blocks; for every choice of S/2 blocks as
    in-sample vs the rest out-of-sample, take the IS-best strategy and record its relative OOS rank ->
    logit. PBO = fraction of splits where the IS winner lands below the OOS median (logit < 0)."""
    T = len(M); N = len(M[0]) if T else 0
    if S % 2 or T < S or N < 2:
        return None
    blk = T // S
    blocks = [list(range(i * blk, T if i == S - 1 else (i + 1) * blk)) for i in range(S)]
    half = S // 2
    n_below = 0; tot = 0
    for combo in itertools.combinations(range(S), half):
        isset = set(combo)
        is_rows = [r for b in combo for r in blocks[b]]
        oos_rows = [r for b in range(S) if b not in isset for r in blocks[b]]
        is_perf = [_mean([M[r][c] for r in is_rows]) for c in range(N)]
        oos_perf = [_mean([M[r][c] for r in oos_rows]) for c in range(N)]
        nstar = max(range(N), key=lambda c: is_perf[c])
        order = sorted(range(N), key=lambda c: oos_perf[c])
        rank = order.index(nstar) + 1
        omega = rank / (N + 1.0)
        lam = math.log(omega / (1.0 - omega))
        n_below += (1 if lam < 0 else 0); tot += 1
    return (n_below / tot) if tot else None


def promotion_gate(dsr, pbo, min_dsr=0.95, max_pbo=0.5):
    """Machine-readable release gate: deployable only if DSR >= min_dsr AND PBO <= max_pbo."""
    okd = (dsr is not None and dsr >= min_dsr)
    okp = (pbo is not None and pbo <= max_pbo)
    reasons = []
    if not okd:
        reasons.append("DSR %.3f < %.2f (selection-adjusted Sharpe not significant)" % (dsr if dsr is not None else float("nan"), min_dsr))
    if not okp:
        reasons.append("PBO %.2f > %.2f (high backtest-overfit probability)" % (pbo if pbo is not None else float("nan"), max_pbo))
    return {"deployable": bool(okd and okp), "dsr": dsr, "pbo": pbo, "reasons": reasons or ["passes DSR + PBO"]}


def gen_fixture():
    # deterministic 24x4 performance matrix: strategy 0 has a genuine per-period edge, others are a fixed
    # pseudo-pattern (no RNG, so Py and JS lock to identical PBO).
    T, N = 24, 4
    M = [[((i * 7 + c * 13) % 11 - 5) / 5.0 + (0.6 if c == 0 else 0.0) for c in range(N)] for i in range(T)]
    pbo = pbo_cscv(M, S=6)
    splits = purged_kfold(20, 4, embargo=2)
    g_ok = promotion_gate(0.99, 0.10)
    g_dsr = promotion_gate(0.80, 0.10)
    g_pbo = promotion_gate(0.99, 0.70)
    return {"fixture_version": 1, "case": "validation-core",
            "M": M, "S": 6, "pbo": pbo,
            "splits": [[tr, te] for tr, te in splits],
            "gateOk": g_ok, "gateDsr": g_dsr, "gatePbo": g_pbo}


def main():
    p = os.path.join(os.path.dirname(os.path.abspath(__file__)), "..", "validation_golden.json")
    json.dump(gen_fixture(), open(p, "w"), separators=(",", ":"))
    print("wrote", os.path.normpath(p))


if __name__ == "__main__":
    main()

File: validation_engine.js
/* validation_engine.js — overfit promotion gate, 1:1 port of tools/market_map/validation_engine.py.
   purgedKfold (embargoed CV splits) + pboCscv (Probability of Backtest Overfitting, CSCV) + promotionGate
   (DSR + PBO). Locked to tools/validation_golden.json by tools/test_validation_parity.mjs. window.MrktValid. */
(function (root) {
  'use strict';
  function purgedKfold(n, k, embargo) {
    embargo = (embargo == null ? 0 : embargo);
    if (k < 2 || n < k) return [];
    var fold = Math.floor(n / k), out = [];
    for (var i = 0; i < k; i++) {
      var lo = i * fold, hi = (i === k - 1) ? n : (i + 1) * fold;
      var test = []; for (var t = lo; t < hi; t++) test.push(t);
      var elo = Math.max(0, lo - embargo), ehi = Math.min(n, hi + embargo);
      var train = []; for (var j = 0; j < n; j++) if (j < elo || j >= ehi) train.push(j);
      out.push([train, test]);
    }
    return out;
  }
  function _mean(xs) { return xs.length ? xs.reduce(function (p, x) { return p + x; }, 0) / xs.length : 0; }
  function _combos(arr, half) {
    var res = [];
    (function rec(start, cur) {
      if (cur.length === half) { res.push(cur.slice()); return; }
      for (var i = start; i < arr.length; i++) { cur.push(arr[i]); rec(i + 1, cur); cur.pop(); }
    })(0, []);
    return res;
  }
  function pboCscv(M, S) {
    S = (S == null ? 8 : S);
    var T = M.length, N = T ? M[0].length : 0;
    if (S % 2 || T < S || N < 2) return null;
    var blk = Math.floor(T / S), blocks = [];
    for (var b = 0; b < S; b++) { var rows = []; var end = (b === S - 1) ? T : (b + 1) * blk; for (var r = b * blk; r < end; r++) rows.push(r); blocks.push(rows); }
    var half = S / 2, idx = []; for (var s = 0; s < S; s++) idx.push(s);
    var combos = _combos(idx, half), nBelow = 0, tot = 0;
    combos.forEach(function (combo) {
      var isset = {}; combo.forEach(function (b2) { isset[b2] = 1; });
      var isRows = [], oosRows = [];
      for (var b2 = 0; b2 < S; b2++) (isset[b2] ? isRows : oosRows).push.apply(isset[b2] ? isRows : oosRows, blocks[b2]);
      var isP = [], oosP = [];
      for (var c = 0; c < N; c++) { isP.push(_mean(isRows.map(function (r) { return M[r][c]; }))); oosP.push(_mean(oosRows.map(function (r) { return M[r][c]; }))); }
      var nstar = 0; for (c = 1; c < N; c++) if (isP[c] > isP[nstar]) nstar = c;
      var order = idxN(N).sort(function (a, b) { return oosP[a] - oosP[b]; });
      var rank = order.indexOf(nstar) + 1, omega = rank / (N + 1);
      var lam = Math.log(omega / (1 - omega));
      if (lam < 0) nBelow++; tot++;
    });
    return tot ? nBelow / tot : null;
  }
  function idxN(N) { var a = []; for (var i = 0; i < N; i++) a.push(i); return a; }
  function promotionGate(dsr, pbo, minDsr, maxPbo) {
    minDsr = (minDsr == null ? 0.95 : minDsr); maxPbo = (maxPbo == null ? 0.5 : maxPbo);
    var okd = (dsr != null && dsr >= minDsr), okp = (pbo != null && pbo <= maxPbo), reasons = [];
    if (!okd) reasons.push('DSR ' + (dsr != null ? dsr.toFixed(3) : 'NaN') + ' < ' + minDsr.toFixed(2) + ' (selection-adjusted Sharpe not significant)');
    if (!okp) reasons.push('PBO ' + (pbo != null ? pbo.toFixed(2) : 'NaN') + ' > ' + maxPbo.toFixed(2) + ' (high backtest-overfit probability)');
    return { deployable: !!(okd && okp), dsr: dsr, pbo: pbo, reasons: reasons.length ? reasons : ['passes DSR + PBO'] };
  }
  var API = { purgedKfold: purgedKfold, pboCscv: pboCscv, promotionGate: promotionGate };
  if (typeof module !== 'undefined' && module.exports) module.exports = API;
  root.MrktValid = API;
})(typeof globalThis !== 'undefined' ? globalThis : this);
